pkg/kubelet/kubelet.go

```diff
@@ -349,12 +369,37 @@ func (self *podWorkers) Run(podFullName string, action func()) {
 	}()
 }
 
-func makeEnvironmentVariables(container *api.Container) []string {
+func (kl *Kubelet) makeEnvironmentVariables(container *api.Container) ([]string, error) {
 	var result []string
+	// Add service environment variables.  Note:  These are added to the docker.Config, but are
+	// not included in the checksum computed by dockertools.BuildDockerName(...).  That way,
+	// we can still determine whether an api.Container is already running by its hash.
+	// (We don't want to restart a container just because some service changed.)
+	//
+	// TODO: limit env vars to same namespace services, plus kubernetes{-ro}. (Issue #3293)
+	//
+	// Note that there is a race between Kubelet seeing the pod and kubelet seeing the service.
+	// To avoid this users can: (1) wait between starting a service and starting; or (2) detect
+	// missing service env var and exit and be restarted; or (3) use DNS instead of env vars
+	// and keep trying to resolve the DNS name of the service (recommended).
+	serviceEnv, err := kl.getServiceEnvVarMap()
+	if err != nil {
+		return result, err
+	}
 	for _, value := range container.Env {
+		// The code is in transition from using etcd+BoundPods to apiserver+Pods.
+		// So, the master may set service env vars, or kubelet may.  In case both are doing
+		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
+		// env vars.
+		// TODO: remove this net line once all platforms use apiserver+Pods.
+		delete(serviceEnv, value.Name)
```

Comment:
In our internal experience, kubelet usually know more about the details of how to implement QoS than the master or the scheduler.  But, if that proves to not be true, the master or scheduler can put a hint to kubelet in an annotation.